src/deepsparse/timing/timing_builder.py

```diff
@@ -0,0 +1,197 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from deepsparse.timing.timing_schema import InferenceTimingSchema
+
+
+__all__ = ["TimingBuilder"]
+
+
+class TimingBuilder:
+    """
+    The goal of the object is to aggregate the durations
+    (time deltas in seconds) of various components of the inference
+    pipeline.
+    Once all the desired time deltas are aggregated, the
+    object may build the data structure that summarizes the
+    collected information.
+
+    The envisioned flow of how the object aggregates time
+    deltas is depicted below
+
+    ```
+    builder.start()
+
+    builder.{process_name}_start()
+    something_happens()
+    builder.{process_name}_complete()
+    ...
+    summary = builder.build()
+    ```
+
+    The object can be extended to aggregate further deltas.
+    To extend, four things are required:
+        - adding {process_name}_delta property (including the setter)
+        - adding {process_name}_start method
+        - adding {process_name}_complete method
+        - adding the {process_name_delta} to the arguments collected
+          inside the build() method.
```

Comment:
thoughts on the following?

1. Removing the specific `_start()` and `_complete()` methods in favor of `builder.start_phase(name)`, `builder.end_phase(name)`.  
2. Keeping `start_times` and `end_times` dicts internally that are populated by `start/end_phase` functions
3. measuring deltas between the dict keys on build, and passing in as kwargs to the timing object

don't think that users need access to the deltas before build and this makes it a bit more flexible.  we can either fail if invalid phase names are added, or pipe them in as extras to the timing object